File: python/hard/sub_array_sort/subArraySort.py
#  O(n) time | O(1) space
def subArraySort(array):
    minOutOfOrder = float('inf')
    maxOutOfOrder = float('-inf')

    for i in range(len(array)):
        num = array[i]
        if isOutOfOrder(i, num, array):
            minOutOfOrder = min(minOutOfOrder, num)
            maxOutOfOrder = max(maxOutOfOrder, num)
    if minOutOfOrder == float('inf'):
        return [-1, -1]

    leftSubArrayIdx = 0
    while minOutOfOrder >= array[leftSubArrayIdx]:
        leftSubArrayIdx += 1

    rightSubArrayIdx = len(array) - 1
    while maxOutOfOrder <= array[rightSubArrayIdx]:
        rightSubArrayIdx -= 1

    return [leftSubArrayIdx, rightSubArrayIdx]


def isOutOfOrder(i, num, array):
    if i == 0:
        return num > array[i+1]
    if i == len(array) - 1:
        return num < array[i-1]
    return num > array[i+1] or num < array[i-1]


print(subArraySort([1, -1]))

File: javascript/hard/sub_array_sort/subArraySort.js
// O(n) time | O(1) space
function subArraySort(array) {
    let minOutOfOrder = Infinity;
    let maxOutOfOrder = -Infinity;

    for (let i = 0; i < array.length; i++) {
        let num = array[i];
        if (isOutOfOrder(i, num, array)) {
            minOutOfOrder = Math.min(minOutOfOrder, num);
            maxOutOfOrder = Math.max(maxOutOfOrder, num);
        }
    }

    if (minOutOfOrder === Infinity) {
        return [-1, -1];
    }
    let leftSubArrayIdx = 0;
    while (minOutOfOrder >= array[leftSubArrayIdx]) {
        leftSubArrayIdx++;
    }
    let rightSubArrayIdx = array.length - 1;
    while (maxOutOfOrder <= array[rightSubArrayIdx]) {
        rightSubArrayIdx -= 1;
    }
    return [leftSubArrayIdx, rightSubArrayIdx];
}

function isOutOfOrder(i, num, array) {
    if (i === 0) {
        return num > array[i + 1];
    }
    if (i === array.length - 1) {
        return num < array[i - 1];
    }
    return num > array[i + 1] || num < array[i - 1];
}

console.log(subArraySort([1, 2, 3, 4, -1, 5, 6, 2, 9]));